### PR TITLE
feat(hotspot): calculator auto detect hotkey in the hot partition

### PR DESCRIPTION
**Pull request by Smityz**
### What problem does this PR solve? 
Add a policy for `hotspot_partition_calculator` to send RPC to start hotkey detect in the replica_server auto.
### What is changed and how it works?
we can set `enable_hotkey_auto_detect = true` to start auto detect hotkey, and set `hot_partition_threshold` and `occurrence_threshold` to adjust sensitivity.
## Config changes
 
```diff
[pegasus.collector]
+ enable_hotkey_auto_detect
+ hot_partition_threshold
+ occurrence_threshold
```

### CODE REVIEW

File: src/server/hotspot_partition_calculator.cpp [hunk 1]
```diff
@@ -128,6 +147,36 @@ void hotspot_partition_calculator::data_analyse()
...
+void hotspot_partition_calculator::detect_hotkey_in_hotpartition(int data_type)
+{
+    for (int index = 0; index < _hot_points.size(); index++) {
+        if (_hot_points[index][data_type].get()->get_value() >= FLAGS_hot_partition_threshold) {
+            if (++_hotpartition_pool[index][data_type] >= FLAGS_occurrence_threshold) {
+                derror_f("Find a {} hot partition {}.{}",
+                         (data_type == partition_qps_type::READ_HOTSPOT_DATA ? "read" : "write"),
+                         _app_name,
+                         index);
+                FAIL_POINT_INJECT_F("send_hotkey_detect_request", [](dsn::string_view) {});
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.cpp:167]: you can move this fail point into func `send_hotkey_detect_request`

File: src/server/hotspot_partition_calculator.h [hunk 2]
```diff
@@ -18,12 +18,15 @@
 #pragma once
 
 #include "hotspot_partition_stat.h"
+#include <dsn/utility/flags.h>
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.h:21]: move `#include <dsn/utility/flags.h>` behind of `#include <gtest/gtest_prod.h>`

File: src/server/test/hotspot_partition_test.cpp [hunk 3]
```diff
@@ -92,6 +107,29 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
     test_rows[HOT_SCENARIO_1_WRITE_HOT_PARTITION].put_qps = 5000.0;
     expect_vector = {{0, 0, 0, 4, 0, 0, 0, 0}, {0, 0, 4, 0, 0, 0, 0, 0}};
     test_policy_in_scenarios(test_rows, expect_vector, calculator);
+    clear_calculator_histories(calculator);
+}
+
+TEST_F(hotspot_partition_test, send_hotkey_detect_request)
+{
+    const int HOT_SCENARIO_0_READ_HOT_PARTITION = 7;
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:115]: what does `HOT_SCENARIO_0` mean? could you remove this prefix?

File: src/server/test/hotspot_partition_test.cpp [hunk 4]
```diff
@@ -19,14 +19,19 @@
...
 namespace server {
 
 class hotspot_partition_test : public pegasus_server_test_base
 {
 public:
-    hotspot_partition_test() : calculator("TEST", 8){};
+    hotspot_partition_test() : calculator("TEST", 8)
+    {
+        dsn::fail::setup();
+        dsn::fail::cfg("send_hotkey_detect_request", "return()");
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:33]: you should call `dsn::fail::teardown()` in the dstor of `hotspot_partition_test`

File: src/server/test/hotspot_partition_test.cpp [hunk 5]
```diff
@@ -58,6 +63,16 @@ class hotspot_partition_test : public pegasus_server_test_base
         std::vector<std::vector<double>> result = get_calculator_result(calculator._hot_points);
         ASSERT_EQ(result, expect_result);
     }
+    void aggregate_analyse_data(std::vector<row_data> scenario,
+                                hotspot_partition_calculator &calculator)
+    {
+        calculator.data_aggregate(std::move(scenario));
+        calculator.data_analyse();
+    }
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:71]: add a blank line between two fuctions.

File: src/server/test/hotspot_partition_test.cpp [hunk 6]
```diff
@@ -58,6 +63,16 @@ class hotspot_partition_test : public pegasus_server_test_base
         std::vector<std::vector<double>> result = get_calculator_result(calculator._hot_points);
         ASSERT_EQ(result, expect_result);
     }
+    void aggregate_analyse_data(std::vector<row_data> scenario,
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:66]: `std::vector<row_data> &scenario`

File: src/server/test/hotspot_partition_test.cpp [hunk 7]
```diff
@@ -92,6 +107,29 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+    test_rows[HOT_SCENARIO_0_WRITE_HOT_PARTITION].put_qps = 5000.0;
+    for (int i = 0; i < FLAGS_occurrence_threshold; i++) {
+        aggregate_analyse_data(test_rows, calculator);
+    }
+    std::vector<std::array<int, 2>> expect_result = {
+        {0, 100}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {100, 0}};
+    ASSERT_EQ(calculator._hotpartition_pool, expect_result);
+
+    test_rows = generate_row_data();
+    for (int i = 0; i < 50; i++) {
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:128]: what's the meaning of  number `50`?

File: src/server/hotspot_partition_calculator.cpp [hunk 8]
```diff
@@ -42,6 +43,24 @@ DSN_DEFINE_int64("pegasus.collector",
                  "eliminate outdated historical "
                  "data");
 
+DSN_DEFINE_bool("pegasus.collector",
+                enable_hotkey_auto_detect,
+                false,
+                "auto detect hot key in the hot paritition");
+
+// if hot_partition_counter >= FLAGS_hotpartition_threshold, This partition is a hot partition
+DSN_DEFINE_int32("pegasus.collector",
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.cpp:52]: I think you can move these comments into the descripton param of `DSN_DEFINE_xxx`

File: src/server/test/hotspot_partition_test.cpp (hunk 6 above)
Comment (Smityz) [reply src/server/test/hotspot_partition_test.cpp:66]: In line 69
```cpp
calculator.data_aggregate(std::move(scenario));
```
It can't be a ref

File: src/server/test/hotspot_partition_test.cpp [hunk 9]
```diff
@@ -92,6 +114,30 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+
+TEST_F(hotspot_partition_test, send_hotkey_detect_request)
+{
+    const int READ_HOT_PARTITION = 7;
+    const int WRITE_HOT_PARTITION = 0;
+    std::vector<row_data> test_rows = generate_row_data();
+    test_rows[READ_HOT_PARTITION].get_qps = 5000.0;
+    test_rows[WRITE_HOT_PARTITION].put_qps = 5000.0;
+    for (int i = 0; i < FLAGS_occurrence_threshold; i++) {
+        aggregate_analyse_data(test_rows, calculator);
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:128]: ` aggregate_analyse_data(generate_row_data(), calculator);`
remove the `test_rows`, so you can reduce the copy count of `std::vector<row_data>` by 1.

File: src/server/hotspot_partition_calculator.h [hunk 10]
```diff
@@ -17,13 +17,17 @@
...
 #include <gtest/gtest_prod.h>
+
 #include <dsn/perf_counter/perf_counter.h>
+#include <dsn/utility/flags.h>
+#include "hotspot_partition_stat.h"
 
 namespace pegasus {
 namespace server {
 
+DSN_DECLARE_int32(occurrence_threshold);
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.h:29]: why did you add `DSN_DECLARE_int32`? I think you can remove it. Add it in where you really use.

File: src/server/test/hotspot_partition_test.cpp [hunk 11]
```diff
@@ -92,6 +114,30 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+    std::vector<std::array<int, 2>> expect_result = {
+        {0, 100}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {100, 0}};
+    ASSERT_EQ(calculator._hotpartition_pool, expect_result);
+
+    test_rows = generate_row_data();
+    const int back_to_normal = 50;
+    for (int i = 0; i < back_to_normal; i++) {
+        aggregate_analyse_data(test_rows, calculator);
+    }
+    expect_result = {{0, 50}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {50, 0}};
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:139]: Does this `50` has relative with `back_to_normal`?

File: src/server/test/hotspot_partition_test.cpp [hunk 12]
```diff
@@ -92,6 +114,30 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+}
+
+TEST_F(hotspot_partition_test, send_hotkey_detect_request)
+{
+    const int READ_HOT_PARTITION = 7;
+    const int WRITE_HOT_PARTITION = 0;
+    std::vector<row_data> test_rows = generate_row_data();
+    test_rows[READ_HOT_PARTITION].get_qps = 5000.0;
+    test_rows[WRITE_HOT_PARTITION].put_qps = 5000.0;
+    for (int i = 0; i < FLAGS_occurrence_threshold; i++) {
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:127]: There are too many same code between 127-132 and 135-140

File: src/server/test/hotspot_partition_test.cpp [hunk 13]
```diff
@@ -92,6 +114,30 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+        aggregate_analyse_data(test_rows, calculator);
+    }
+    std::vector<std::array<int, 2>> expect_result = {
+        {0, 100}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {100, 0}};
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:131]: Does the 100 has relative with the default value of ` FLAGS_occurrence_threshold`? If so, you should change it to ` FLAGS_occurrence_threshold`

File: src/server/hotspot_partition_calculator.cpp [hunk 14]
```diff
@@ -42,6 +43,24 @@ DSN_DEFINE_int64("pegasus.collector",
                  "eliminate outdated historical "
                  "data");
 
+DSN_DEFINE_bool("pegasus.collector",
+                enable_hotkey_auto_detect,
```
Comment (acelyc111) [inline src/server/hotspot_partition_calculator.cpp:47]:
```suggestion
                enable_hotkey_detect,
```

File: src/server/test/config.ini [hunk 15]
```diff
@@ -509,3 +509,6 @@ onebox2 = 2
 [pegasus.clusters]
 onebox = 0.0.0.0:34701
 onebox2 = 0.0.0.0:35701
+
+[pegasus.collector]
+enable_hotkey_auto_detect = true
```
Comment (acelyc111) [inline src/server/test/config.ini:514]:
```suggestion
enable_hotkey_detect = true
```

File: src/server/hotspot_partition_calculator.h [hunk 16]
```diff
@@ -63,7 +66,10 @@ class hotspot_partition_calculator
     // saving historical data can improve accuracy
     stat_histories _partitions_stat_histories;
 
+    std::vector<std::array<int, 2>> _hotpartition_pool;
```
Comment (hycdong) [inline src/server/hotspot_partition_calculator.h:69]: Add comments for this structure to explain the meaning of `array<int, 2>`

File: src/server/hotspot_partition_calculator.cpp [hunk 17]
```diff
@@ -128,15 +145,44 @@ void hotspot_partition_calculator::data_analyse()
...
+                send_hotkey_detect_request(_app_name,
+                                           index,
+                                           (data_type == dsn::apps::hotkey_type::type::READ)
+                                               ? dsn::apps::hotkey_type::type::READ
+                                               : dsn::apps::hotkey_type::type::WRITE,
+                                           dsn::apps::hotkey_detect_action::type::START);
+            }
+        } else {
+            _hotpartition_pool[index][data_type] =
+                std::max(_hotpartition_pool[index][data_type] - 1, 0);
```
Comment (hycdong) [inline src/server/hotspot_partition_calculator.cpp:174]: Why try to decrease the value of `_hotpartition_pool[index][data_type]`? Actually, I don't know the usage of `_hotpartition_pool`.

Comment (Smityz) [reply src/server/hotspot_partition_calculator.cpp]: I add some comments in the codes

**Review by acelyc111 — DISMISSED** [review]

File: src/server/test/hotspot_partition_test.cpp [hunk 18]
```diff
@@ -92,6 +121,37 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
...
+    const int back_to_normal = 30;
+    hotpartition_count = FLAGS_occurrence_threshold - back_to_normal;
+    expect_result = {{0, hotpartition_count},
+                     {0, 0},
+                     {0, 0},
+                     {0, 0},
+                     {0, 0},
+                     {0, 0},
+                     {0, 0},
+                     {hotpartition_count, 0}};
```
Comment (hycdong) [inline src/server/test/hotspot_partition_test.cpp:153]: `expect_result` is set on Line 142, why set the same value again?

Comment (Smityz) [reply src/server/test/hotspot_partition_test.cpp:153]: It is a different test case, expext result changed

**Review by hycdong — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]
